adding Martas models

Marta-Rudnicka opened the pull request.

Commit:
adding Martas models

TJGorrie commented on xchem_db/models.py:
```diff
@@ -1,16 +1,172 @@
+
+#INVENTORY DATA - mostly SoakDB-only classes
+
+class Protein(models.Model):
```
This looks class looks okay

TJGorrie replied:
Implicitly linked to Crystal Table via crystal__visit__proposal__protein

TJGorrie commented on xchem_db/models.py:
```diff
@@ -1,16 +1,172 @@
+        return tpsa(self.smiles)
+
+    def __str__ (self):
+        return self.code
```
Would this stuff be better suited as fields in the Compound table instead of a method that is generated on the fly. Such that the information is always accessible to subsequent applets that use XCDB. The information shouldn't change for a given smiles string so it seems computationally inefficient to generate with a backend call when it can be calculate on data creation (and require a python environment with RDKit).?

TJGorrie commented on xchem_db/models.py:
```diff
@@ -1,16 +1,172 @@
+        return mol_wt(self.smiles)
+    
+    def tpsa(self):
+        return tpsa(self.smiles)
```
Marta says these lines are not meant to be here

reskyner commented:
**Do not merge this** until the issues below have been addressed, and the CI tests pass. As far as I can see we need to:

~~- [ ] Add rdkit as a requirement~~
~~- [ ] Specify how to install rdkit in the relevant actions (.actions/<whatever>)~~
- [ ] Check that building the database from the sql statements works
- [ ] Create a new example of data in data.json that covers the new models
- [ ] Test that we can populate the database with sql statements (i.e. ``python manage.py loaddata data.json``)

TJGorrie commented on xchem_db/models.py:
```diff
@@ -1,16 +1,172 @@
+    it is brought in by the user'''
+
+    name = models.CharField(max_length=100)
+    for_industry = models.BooleanField(default=False)
+    public = models.BooleanField(default=False)
```
Is this a suitable method of hiding industry data? How would we stop a naive PhD student from looking at data that is supposed to be kept hidden? 

Perhaps for frontend facing application the authentication shake-down showing only the data people are supposed to see. But from the perspective of the other use-cases of XCDB is this suitable enough or is this a different problem all together @reskyner

TJGorrie commented on xchem_db/models.py:
```diff
@@ -1,16 +1,172 @@
+    name = models.CharField(max_length=100)
+    library = models.ForeignKey(Library, on_delete=models.PROTECT, related_name="plates" )
+    current = models.BooleanField(default=True)
+    last_tested =  models.DateField(auto_now=True)
```
Make a note of when and how this should get updated

TJGorrie commented on xchem_db/models.py:
```diff
@@ -1,16 +1,172 @@
+    library_plate =  models.ForeignKey(LibraryPlate, on_delete=models.CASCADE, related_name="compounds")
+    well  = models.CharField(max_length=4)
+    concentration = models.IntegerField(null=True, blank=True)
+
```
Not Important: but do we maybe need fields to SourceWell or Library plate to indicate additional things: e.g. Buffer/Solvent Contents, Well Volumes, pH?

reskyner replied:
What information is currently captured about an experiment that isn't recorded in soakDB?

TJGorrie commented on xchem_db/models.py:
```diff
@@ -1,16 +1,172 @@
+    class Meta:
+        if os.getcwd() != '/dls/science/groups/i04-1/software/luigi_pipeline/pipelineDEV':
+            app_label = 'xchem_db'
+        db_table = 'preset'
```
library_preset ?

TJGorrie commented on xchem_db/models.py:
```diff
@@ -21,46 +177,44 @@ class Meta:
-    proposal = models.CharField(max_length=255, blank=False, null=False, unique=True, db_index=True)
+
+    # TODO - can we refactor this for title [original comment]
+    name = models.CharField(max_length=255, blank=False, null=False, unique=True)
```
Maybe change back to proposal, I don't know who this affects. We should be able to change pipeline code to accommodate no biggie

reskyner replied:
leave as proposal

TJGorrie commented on xchem_db/models.py:
```diff
@@ -102,11 +268,122 @@ class Crystal(models.Model):
+
+
+#new class (SoakDB experimental data)	
+class SoakDBCompound(models.Model):
+    '''Compound data copied from inventory data when the compound is used
+    in the experiment'''
+
+    proposal = models.ForeignKey(Proposals, related_name="exp_compounds", on_delete=models.CASCADE, blank=True, null=True)
+    library_name = models.CharField(max_length=100)
+    library_plate = models.CharField(max_length=100)
+    well = models.CharField(max_length=4)
+    code = models.CharField(max_length=100)
+    smiles = models.CharField(max_length=256)
+    crystal = models.ForeignKey(Crystal, related_name="compounds", on_delete=models.PROTECT, blank=True, null=True) #to allow cocktails
+    
+    class Meta:
+        if os.getcwd() != '/dls/science/groups/i04-1/software/luigi_pipeline/pipelineDEV':
+            app_label = 'xchem_db'
+        db_table = 'soakdb_compound'
+
```
We 'XChem' need to have an extremely low-level Frank-less discussion about how we should implement combi soak crystals.

TJGorrie commented on xchem_db/models.py:
```diff
@@ -102,11 +268,122 @@ class Crystal(models.Model):
+    cryo_location = models.CharField(max_length=4, blank=True, null=True)
+
+    #based on macros in the original SoakDB file; consulted with Ailsa
+    def soak_vol(self): #supposing it means soak transfer volume -- needs verification
```
would 
```
def soak_vol(self, min_vol_unit):
         volume = (self.crystal_plate.drop_volume * self.solv_frac) / (100 - self.solv_frac)
         return round(volume /min_vol_unit , 0) * min_vol_unit
```
be less 'black-boxy'.

reskyner replied:
method removed

TJGorrie commented on xchem_db/models.py:
```diff
@@ -102,11 +268,122 @@ class Crystal(models.Model):
+    cryo_frac = models.FloatField(blank=True, null=True)
+    cryo_stock_frac = models.FloatField(blank=True, null=True)
+    cryo_location = models.CharField(max_length=4, blank=True, null=True)
+
```
Or 
min_vol_unit = models.FloatField(default=2.5)?

reskyner replied:
method removed

reskyner commented:
On second thought @Marta-Rudnicka - no rdkit in here please. 

Add the keys in your ``properties`` dict as actual fields in the model, with appropriate data-types. You can calculate and populate these values from your own code when you're inserting data into the database.

TJGorrie commented on xchem_db/models.py:
```diff
@@ -175,32 +451,53 @@ class Meta:
 
-
+#heavily modified; some attributes added, some moved to Batch
 class Lab(models.Model):
```
Looks okay I think.

TJGorrie commented on xchem_db/models.py:
```diff
@@ -175,32 +451,53 @@ class Meta:
             app_label = 'xchem_db'
         db_table = 'lab'
 
+'''
+#(requires installing json field:
+#pip install jsonfield)
+from jsonfield import JSONField
+
+class SolventAutofill(models.Model):
+    #patterns of typical parameters used in solvent testing; used for auto-filling data
+    
+    name = models.CharField(max_length=100)
+    solvents = JSONField() #array of strings
+    solvent_concentrations = JSONField() #array of floats
+    cryo = models.BooleanField()
+    comment = models.textField()
+'''
+
+
+#ONLY OLD UNCHANGED MODELS BELOW
```
Add jsonfield to requirements on xchemdb and Add the class.

TJGorrie replied:
Or figure out what is happening when XCDB migrates.

reskyner replied:
Array fields specific to Postgres - need to work through json field stuff?

TJGorrie reviewed (approved):
Everything looks okay. I would like to schedule a meeting with Alice + Daren so we can really narrow down how we are supposed to communicate combs-soaked fragments and HOW this information will be transferred and communicated downstream in both XCR and Fragalysis. These will be single crystal (.pdb) files with multiple compounds. How do we handle this in a programmatic sense. 

Currently, there are physical hacks that are being done which are kludges at best, I would prefer to nail the implementation down now before it becomes a huge headache and before Frank decides to change the requirements too.

Marta-Rudnicka committed:
removed rdkit and "if os.getcwd()" statements

reskyner commented:
Still failing CI - please check the logs for the CI job

Marta-Rudnicka committed:
made Lab.compound nullable

Marta-Rudnicka committed:
nuked old db and added SPA models

Marta-Rudnicka commented:
I deleted `init.sql`, `db.json` and migrations, and I added the latest iteration of my `models.py`. Now the tests fail at the 'setup environment' stage:
`E: Package 'python-psycopg2' has no installation candidate`
`Error: Process completed with exit code 100.`

It doesn't look like anything to do with my models, and I don't know how to debug it. @reskyner @TJGorrie

reskyner committed:
Merge pull request #9 from xchem/main

Update actions apt installs

reskyner committed:
Auto-commit of new versions of migrations, SQL files & schema image
